pkg/execution/pauses/block.go

```diff
@@ -0,0 +1,464 @@
+}
+
+// Compact reads all indexed deletes from block for an index, then compacts any blocks over a given threshold
+// by removing pauses and rewriting blocks.
+func (b *blockstore) Compact(ctx context.Context, idx Index) {
+	// Implement the following:
+
+	// TODO: Lease compaction for the index.
+	// TODO: Read all block metadata for the index
+	// TODO: Read all blockDeleteKey entries for the index
+	// TODO: For each deleted entry, record which block the delete is for.
+	//       XXX: we can't embed the timestamp in block IDs because of old parallelism;  how do we track this?
+	// TODO: If len(block_deletes) > block_compact_ratio, rewrite the block by:
+	// 1. fetching the block
+	// 2. filtering deleted pauses from the block
+	// 3. rewriting the block
+}
+
+func (b *blockstore) blockIDForTimestamp(ctx context.Context, idx Index, ts time.Time) (*ulid.ULID, error) {
+	score := strconv.Itoa(int(ts.UnixMilli()))
+	ids, err := b.rc.Do(
+		ctx,
+		b.rc.B().Zrange().Key(b.blockIndexKey(idx)).Min("("+score).Max("+inf").Limit(0, 1).Build(),
+	).AsStrSlice()
+	if len(ids) == 1 {
+		id, err := ulid.Parse(ids[0])
+		return &id, err
+	}
+	if err == nil || rueidis.IsRedisNil(err) {
+		return nil, nil
+	}
+	return nil, err
+}
+
+func (b *blockstore) blockMetadata(ctx context.Context, idx Index, block *Block) (*blockMetadata, error) {
+	earliest, err := b.buf.PauseTimestamp(ctx, idx, *block.Pauses[0])
+	if err != nil {
+		return nil, fmt.Errorf("error fetching earliest pause time: %w", err)
+	}
+	latest, err := b.buf.PauseTimestamp(ctx, idx, *block.Pauses[len(block.Pauses)-1])
+	if err != nil {
+		return nil, fmt.Errorf("error fetching latest pause time: %w", err)
+	}
+
+	// Block indexes are a zset of blocks stored by last pause timestamp,
+	// which is embedded into the pause ID.
+	//
+	// We also have a mapping of block ID -> metadata, storing the timeranges and
+	// current block size.  This is used during compaction.
+	return &blockMetadata{
+		Timeranges: [2]int64{earliest.UnixMilli(), latest.UnixMilli()}, // earliest/latest
+		UUIDranges: [2]uuid.UUID{block.Pauses[0].ID, block.Pauses[len(block.Pauses)-1].ID},
+		Len:        len(block.Pauses),
+	}, nil
+}
+
+func (b *blockstore) addBlockIndex(ctx context.Context, idx Index, block *Block, md *blockMetadata) error {
+	// Block indexes are a zset of blocks stored by last pause timestamp,
+	// which is embedded into the pause ID.
+	//
+	// We also have a mapping of block ID -> metadata, storing the timeranges and
+	// current block size.  This is used during compaction.
+	metadata, err := json.Marshal(md)
+	if err != nil {
+		return err
+	}
+
+	cmd := b.rc.B().
+		Zadd().
+		Key(b.blockIndexKey(idx)).
+		ScoreMember().
+		ScoreMember(
+			float64(ulid.Time(block.ID.Time()).UnixMilli()),
+			block.ID.String(),
+		).
+		Build()
+	if err := b.rc.Do(ctx, cmd).Error(); err != nil {
+		return err
+	}
+
+	return b.rc.Do(
+		ctx,
+		b.rc.B().
+			Hset().
+			Key(b.blockMetadataKey(idx)).
+			FieldValue().
```

Comment:
should these be transactions?
what do we do when we have failures?

Reply:
This is idempotent.  Writing the same block twice will cause this to rerun.

Reply:
We also _need_ this to be idempotent because we have already written the entire block to storage, and if this tx fails to commit we're in the same place.  So, it's optional to put this in a lua script and won't impact us.